packages/dds/tree/api-report/tree.api.md

```diff
@@ -2024,18 +2029,25 @@ export type TreeValue<TSchema extends ValueSchema = ValueSchema> = [
 
 // @public
 export interface TreeView<TSchema extends ImplicitFieldSchema> extends IDisposable {
-    readonly error?: SchemaIncompatible;
+    readonly compatibility: SchemaCompatibilityStatus;
     readonly events: ISubscribable<TreeViewEvents>;
+    initialize(content: InsertableTreeFieldFromImplicitField<TSchema>): void;
     get root(): TreeFieldFromImplicitField<TSchema>;
     set root(newRoot: InsertableTreeFieldFromImplicitField<TSchema>);
     upgradeSchema(): void;
 }
 
+// @public
+export interface TreeViewConfiguration<TSchema extends ImplicitFieldSchema = ImplicitFieldSchema> {
```

Comment:
Tree configuration was a class to allow for more ergonomic strong typing when declaring instances of it. For example

`const thing = new Foo(...);`  gets strong typing'.

`const foo = {....}` does not get type errors if the definition is wrong.

`const foo: Foo = {....}` loses schema aware typing.

`const foo: Foo = {....} satisfies Foo` is almost as good as the class, but less safe checking of generic type paramaters and requires covariance.

Reply:
Thus I think you might want to continue with that pattern, or remove this type and just use the ImplicitFieldSchema directly.

Reply:
Using a class also allows us to run validation in the constructor and give nice errors.

I haven't found examples of how this configuration is uses yet, so the above is assuming its something like how TreeConfiguration was used.

Reply:
I hadn't considered the ergonomics of the typing. I'm not too opinionated here, but wanted named arguments to `viewWith` (the app-defined metadata is a planned extension, and it looks like Alejandro has a PR out which may want to add one defining policy around schema validation for inserted trees), at which point it seemed like the class wasn't serving much of a purpose.

Reply:
If we want the strong typing, what's actually needed is forcing through a factory function. Using a class and constructor is an intuitive way to do that (has good discoverability). I think doing that, even when named arguments, is worth it for any case where we want to create something that captures schema aware information that people may want to save in a `const`.

For an example containerSchema doesn't take the class approach, and that lead to a lot of use of it not actually being type safe.

I however don't like extra unneeded classes and prefer interfaces for input types like this, so there are incentives both ways. With more complete/realistic examples I think I'd favor classes, but its clearly subjective.

Reply:
Is the typing we're losing with all the alternatives to the class specifically related to the tree schema and initial tree because of its complexities and dynamic nature? Because if that's the case, a separate property bag interface for the simple flags/configuration could be an option for what I'm doing (and apply to `viewWith()`), right? Even a "bare object" in that case [runs into pretty dev-friendly compile-time issues in the call to `schematize()` in the current version of the API](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgJIGUEAsIFs5jABeEAClAPYAOAzsgN4BQyLyEIcARgDYQBqcbsAAmBYBRAAuZJwoVecEAG5GAX0aMYAVxAJCE5DWx4xJABRGc+AIIhhqEMEKCAKlAgRpigJ4AaZAAe0hjG+IQk5NQ0AJQM6owIEjRgyDByyAC8DGwcPPyCImIS0mBQWiiqKokgyTJwUMGYVqZklLSZ2excvAJCovpSyKXlyJUJSSmccEQd9Dnd+X1FgwBEoABuBcIro1UTdUQATLPzeb2FAwD6hyVlFSqMlibhEGZzAPQAVE82dg5OwFc7ggn3eo38aQo0RUPxab2QX1htnsjmc3DcHlB4LqUGhj1CcI+3wJyP+aIxILBqn8UyIeNhL3hiJJf1RgPRwKx1IOh2hQA):
![image](https://github.com/microsoft/FluidFramework/assets/716334/84e96093-760a-4662-b28d-dedcd41ca8c9)

Reply:
Although if even in `viewWith()` world we're going to keep the class for ergonomics, I think I agree with Craig that a single property bag passed to its constructor (with `schema` and other properties like the one I'm adding) seems preferrable.

Reply:
Seems reasonable to me. Latest version of this PR takes that approach with `TreeViewConfiguration`.